### ISSUE TITLE
ask for some training tips on custom dataset.

### ISSUE THREAD

**Issue by wcyjerry** [ultralytics/ultralytics#5032]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

Hi, 
    Sry to bother you guys, but it would be of great help if you could give me some suggestions on my custom task.
Generally, the model usage is to alert if some vehicles or person getting close, and it needs super super fast on devices. the imgsz for training is set to 320, and current model is yolov8n and post-quantize it with int8, this barely meets the fps needs.
    From my understanding on my task, some remote objects is out of my concern and no need to detect, and the model size and input imgsz is not allow me to detect remote(small) object.
My main idea is to make this task easy, I just need it to learn big and obvious object, if it can detect small object, I don't want to punish it, But if not, I don't want to punish it too.
   But my data, is full labeled with almost all objects even it is remote and small. I have two strategies dealing with data:
        1) just train it, and from the model's prediction results, manually filter out results with an area smaller than a threshold.
        2) just remove the labels which area is under an thresh, and train.
    In the first way, I'm not sure whether the small object labels will make it too hard to learn and make bad effects.
    In the second way, I'm afraid I make the task more complex, because maybe, it predict a small object, but I say it is not a positive prediction. I afraid model may get confused, or it will need learn how to distinguish an object is remote or close which is contrary to my initial idea of making the task simpler.
    Is there any advice on my task you can provide? Thanks for your valuable time.

**Comment by glenn-jocher** [comment]
@wcyjerry hello,

First of all, no need to apologize, we are here to help! Your question is a perfectly valid one and it's great that you are thinking critically about your task.

From your explanation, it sounds like you are mainly interested in detecting relatively large or close-by objects with the YOLOv8n model, while ignoring or deemphasizing smaller or distant objects. 

Your proposed solution of training the model with full dataset (including small, remote objects) and manually filtering predicted results may work. However, as you pointed out, the small object labels might bias the model towards detecting those objects it doesn't need to, indirectly affecting detection performance.

Your second solution to pre-process your dataset and remove small labels is also good, but this could potentially confuse the model during training. If it predicts a small object, but there isn't any positive prediction - as you've noted, this increases the complexity rather than simplifying your task.

In this case, you may want to consider revising labeling strategy - for instance, developing a way to differentiate close or large objects from distant or small objects within the labeling process itself. Such differentiation could be done according to size, position or other distinguishable features of the objects of interest.

Additionally, optimizing your model towards your specific needs can also be a handy strategy. This may involve adjusting hyperparameters or feature extractor settings to help your model better recognize the types of objects you are interested in.

Remember, the goal here is to simplify your task. If your task is only concerned with large, obvious objects, your model should ideally focus on learning and detecting those. 

I hope this gives you a better direction for your task.

**Comment by wcyjerry** [comment]
Thx, as you mentioned revising the label strategy, maybe I find a better way to distinguish small object and big object, &nbsp;but I still need to remove labels for small object, seems like no difference between second method I mentioned just, or I get your idea wrong?
In addition do you think it worth a try to modified the loss function like weighted every positive objects according to its box size？In this way small object may conduct small loss, so whether predict it or not, the model get no punish?&nbsp;
And from the exp result, the second method successfully reduce much more false positive &nbsp;
in complex scens, and only predicts big object.

So far it meets my needs, but I will try better ways to train it. And share my experience.

------------------ Original ------------------
From: Glenn Jocher ***@***.***&gt;
Date: Fri,Sep 22,2023 10:42 PM
To: ultralytics/ultralytics ***@***.***&gt;
Cc: Chenyi Wang ***@***.***&gt;, Mention ***@***.***&gt;
Subject: Re: [ultralytics/ultralytics] ask for some training tips on customdataset. (Issue #5032)





 
@wcyjerry hello,
 
First of all, no need to apologize, we are here to help! Your question is a perfectly valid one and it's great that you are thinking critically about your task.
 
From your explanation, it sounds like you are mainly interested in detecting relatively large or close-by objects with the YOLOv8n model, while ignoring or deemphasizing smaller or distant objects.
 
Your proposed solution of training the model with full dataset (including small, remote objects) and manually filtering predicted results may work. However, as you pointed out, the small object labels might bias the model towards detecting those objects it doesn't need to, indirectly affecting detection performance.
 
Your second solution to pre-process your dataset and remove small labels is also good, but this could potentially confuse the model during training. If it predicts a small object, but there isn't any positive prediction - as you've noted, this increases the complexity rather than simplifying your task.
 
In this case, you may want to consider revising labeling strategy - for instance, developing a way to differentiate close or large objects from distant or small objects within the labeling process itself. Such differentiation could be done according to size, position or other distinguishable features of the objects of interest.
 
Additionally, optimizing your model towards your specific needs can also be a handy strategy. This may involve adjusting hyperparameters or feature extractor settings to help your model better recognize the types of objects you are interested in.
 
Remember, the goal here is to simplify your task. If your task is only concerned with large, obvious objects, your model should ideally focus on learning and detecting those.
 
I hope this gives you a better direction for your task.

**Comment by glenn-jocher** [comment]
Hi @wcyjerry,

I'm glad to hear that the second method is already yielding promising results for your specific needs. Indeed, reducing the false positives in complex scenes and focusing on larger objects is a valuable outcome.

As per your new questions, considering the size of the bounding box in the loss function could potentially be a good strategy. This can indeed help in creating a balance between small and large objects and might prevent the model from being penalized heavily for missing small objects. However, be cautious as this approach could also potentially introduce a new source of complexity.

In terms of your question about label strategy, the suggestion was to label objects differently based on their size or other properties, in such a way that makes it easier for model training. However, you're correct in noticing that it is similar to the second method in that it still involves removing or de-emphasizing certain labels.

In the end, the best approach largely depends on the specifics of your project and the characteristics of your data. By iterating and experimenting, you will likely pinpoint more and more what works for your specific scenario.

Please keep us posted about your progress and feel free to ask if you have more questions. Happy training!

**Comment by wcyjerry** [comment]
you mean label different,But how? I'm not very clear about it. could you explain it more?





------------------ Original ------------------
From: Glenn Jocher ***@***.***&gt;
Date: Sat,Sep 23,2023 5:26 AM
To: ultralytics/ultralytics ***@***.***&gt;
Cc: Chenyi Wang ***@***.***&gt;, Mention ***@***.***&gt;
Subject: Re: [ultralytics/ultralytics] ask for some training tips on customdataset. (Issue #5032)





 
Hi @wcyjerry,
 
I'm glad to hear that the second method is already yielding promising results for your specific needs. Indeed, reducing the false positives in complex scenes and focusing on larger objects is a valuable outcome.
 
As per your new questions, considering the size of the bounding box in the loss function could potentially be a good strategy. This can indeed help in creating a balance between small and large objects and might prevent the model from being penalized heavily for missing small objects. However, be cautious as this approach could also potentially introduce a new source of complexity.
 
In terms of your question about label strategy, the suggestion was to label objects differently based on their size or other properties, in such a way that makes it easier for model training. However, you're correct in noticing that it is similar to the second method in that it still involves removing or de-emphasizing certain labels.
 
In the end, the best approach largely depends on the specifics of your project and the characteristics of your data. By iterating and experimenting, you will likely pinpoint more and more what works for your specific scenario.
 
Please keep us posted about your progress and feel free to ask if you have more questions. Happy training!
